rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,506 @@
+Namely, when users create
+[tf.keras.layers.LSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM)
+with `return_sequences=True`, the rest works the same way as CNN.
+
+Unfortunately, this technique does not cover sequence-to-sequence RNN models.
+In these models, the value is the states of encoder, and the query is the input
+of the decoder. The decoder needs to slide its input based on the timesteps, and
+feed them one by one. So, the output of the attention layer at timestep T
+affects the output at T+1.
+
+## Previous Work
+
+[tf.contrib.seq2seq.LuongAttention](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/LuongAttention)
+and
+[tf.contrib.seq2seq.BahdanauAttention](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BahdanauAttention)
+are implementations of dot-product (Luong) and additive (Bahdanau) Attention
+respectively for RNN in Tensorflow. This proposal is based on this
+implementation, but works with CNN/Dense networks.
+
+There is an implementation of Attention as a `tf.layers.Layer` subclass under
+https://github.com/tensorflow/models/tree/master/official/transformer,
+specifically in
+https://github.com/tensorflow/models/blob/master/official/transformer/model/attention_layer.py.
+That file implements dot-product attention proposed in this file, and also
+supports multi-head. Our proposal is to expose such a method in
+`tf.keras.layers`. In addition, our proposal creates variables inside the
+`build()` method, rather than the `call()` method.
+
+There is ongoing work to add Attention in Keras, namely
+https://github.com/keras-team/keras/pull/11421. That proposal addresses
+Attention mechanism for RNN networks only. I cannot see a way to make it work
+for CNN/Dense networks, which are the motivation for our proposal.
+
+https://github.com/keras-team/keras/issues/9263 contains an example of a Keras
+Layer that implements a CNN+Attention network. That example merges CNN and
+Attention into the same class, whereas our proposal is modular. In the
+Examples section, we present an example of how to build a
+CNN+Attention model.
+
+https://github.com/keras-team/keras/issues/7341 is a request to add an Attention
+Layer. Our proposal will resolve that request.
+
+https://github.com/keras-team/keras/issues/7803 is a request for a Multi-Head
+Attention Layer. Multi-Head Attention is not covered in this proposal, but can
+be implemented as a follow-up, as discussed in the
+[Multi-Head Attention](#multi-head-attention) section.
+
+There are a few more issues that request Attention for RNN. They are covered
+either by https://github.com/keras-team/keras/pull/11421 or our proposal:
+
+*   https://github.com/keras-team/keras/issues/5738
+*   https://github.com/keras-team/keras/issues/4962
+*   https://github.com/keras-team/keras/issues/2525
+
+## Design Proposal
+
+We propose to implement the following common attention layers:
+
+*   `Attention`: Basic dot-product attention, a.k.a. Luong-style attention.
```

Comment:
This section should put the equations for both forms, for clarity.

Reply:
Thanks. I put pseudocode in the pydoc in the following section. Let me add that here, too (once I figure out  how to do this in github, sorry for the delay).